Update overview docs on lifetime scopes (#141)
autofac/Documentation: docs/lifetime/index.rst

.. image:: lifetime-scope-tree.png

When you build an Autofac container, what you're creating is that root container/lifetime scope. :doc:`Application integration packages <../integration/index>` or application code can create child lifetime scopes from the container, or even create children from other children.

Lifetime scopes help determine where dependencies come from. *In general*, a component will try to get its dependencies from the lifetime scope resolving the component. For example, if you're in one of the child lifetime scopes and try to resolve something, Autofac will try to get all of the component's dependencies from the child scope.

What affects this mechanism is the "lifetime" aspect of "lifetime scope." Some components, like singletons need to be shared across multiple scopes. This affects how dependencies get located. The "basic rules" are:

- A child lifetime scope can get dependencies from parent scopes, but a parent scope may not reach down into a child scope. (You can locate things by moving "up the tree" but you can't move "down the tree.")

> - A component will get its dependencies from the scope that owns the component even if the component is resolved by a scope further down the tree. For example, singletons are owned by the root lifetime scope because they need to be shared across all child scopes. When you resolve a singleton, *all of its dependencies will come from the root scope*. It doesn't matter if you're resolving it from a child scope and have overrides or other things you think the singleton should use. Doing it this way ensures you can't dispose dependencies out from under the singleton; and that you won't create a memory leak by holding references to things in child scopes after the scope is disposed.

Reviewer: As the diagram, I think it would be useful to be clear that singletons are owned by the lifetime scope they are declared in, not just kept in the root container.

Author: I expanded the singleton discussion into a whole section with code and pictures. That definitely has been a point of confusion, so giving that particular concept some extra attention seemed reasonable.

- Each request is a logical "unit of work" and each requires its own order processing service.
- Each controller needs to log information to the logging service.

In this scenario, you'd have a root lifetime scope that contains the singleton logging service and you'd have one child lifetime scope per request, each with its own controller instance.

From a class registration standpoint, it might look like this:

.. sourcecode:: csharp

    var builder = new ContainerBuilder();

    // The logger is a singleton - shared everywhere.
    builder.RegisterType<Logger>()
           .As<ILogger>()
           .SingleInstance();

    // Controllers are created per lifetime scope. They
    // take an ILogger as a parameter.
    builder.RegisterType<Controller>()
           .InstancePerLifetimeScope();

    var container = builder.Build();

As web requests come in, lifetime scopes might look like this:

.. image:: lifetime-scope-web-example.png

A rough sequence of events for a web app framework like :doc:`ASP.NET Core <../integration/aspnetcore>` goes like this:

1. When a web request comes in, the web application framework creates a child lifetime scope - the "request lifetime scope."

> 2. The web application framework resolves the controller instance from the request lifetime scope. Since the controller is registered as "instance per lifetime scope," it will be a sort of "singleton" within that request but will not be shared with other requests.

Reviewer: It may be a bit risky to mix the use of the word "singleton" in this section, even if it is couched with "sort of". 

(also, comma misplaced)

```suggestion
2. The web application framework resolves the controller instance from the request lifetime scope. Since the controller is registered as "instance per lifetime scope", this instance will be shared by any components within that request but will not be shared with other requests.
```

Author: Updated phrasing, but I think the comma thing is [US English vs. UK English](https://www.grammarly.com/blog/quotation-marks/) - I'll stand by the placement on that one.

Reviewer: US English has commas inside the quote? You guys are bonkers.

Author: You have _no idea_. Oh, wait, you've probably watched politics for the last few years, you _totally have an idea_. 🤪

Reviewer: This is the last straw; it was always going to be something to do with grammar.

.. sourcecode:: csharp

    var builder = new ContainerBuilder();

    // A singleton declared at container build
    // time will be owned by the root scope and get
    // all dependencies from the root scope.
    builder.RegisterType<Component>()
           .SingleInstance();

    // Anything that resolves a Dependency from
    // the root lifetime scope will see the name
    // as 'root'
    builder.Register(ctx => new Dependency("root"));

    var container = builder.Build();

    // Resolve the component from the root lifetime scope.
    var rootComp = container.Resolve<Component>();

    // This will show "root"
    Console.WriteLine(rootComp.Name);

    // Even if we override the Dependency in a child
    // scope, it'll still show "root"
    using (var child1 = container.BeginLifetimeScope(
      b => b.Register(ctx => new Dependency("child1")))
    {
      child1Comp = child1.Resolve<Component>();
      Console.WriteLine(child1Comp.Name);
    }

    // You can override the singleton by adding a new
    // singleton in a child scope. This one will be
    // owned by the child and will work in its children,
    // but it doesn't override at the root scope.

>     using (var child2 = container.BeginLifetimeScope(

Reviewer: Does this actually override the `Component` though? You're only redeclaring the `Dependency` here, but not declaring a new `Component`. Unless I'm missing something obvious.

Author: Dammit, I forgot to redeclare. My brain nowadays. Good catch.

Author: OK, got it.
